File: implementaciones/java/src/Perceptron.java
public class Perceptron {
    public enum Activation {
        LINEAL, ESCALON, SIGMOIDAL, RELU, SOFTMAX, TANH
    }

    private final double learningRate;
    private final int epochs;
    private final Activation activation;
    private double[] weights;
    private double bias;

    public Perceptron(double learningRate, int epochs, Activation activation) {
        if (epochs < 10) {
            throw new IllegalArgumentException("epochs must be >= 10");
        }
        this.learningRate = learningRate;
        this.epochs = epochs;
        this.activation = activation;
    }

    public void train(double[][] x, int[] y) {
        int samples = x.length;
        int features = x[0].length;
        this.weights = new double[features];
        this.bias = 0.0;

        for (int epoch = 0; epoch < epochs; epoch++) {
            for (int i = 0; i < samples; i++) {
                double linear = dot(x[i], weights) + bias;
                double pred = activate(linear);
                double error = y[i] - pred;
                double adjust = activation == Activation.ESCALON
                    ? learningRate * error
                    : learningRate * error * derivative(linear);

                for (int j = 0; j < features; j++) {
                    weights[j] += adjust * x[i][j];
                }
                bias += adjust;
            }
        }
    }

    public int predict(double[] input) {
        double linear = dot(input, weights) + bias;
        double raw = activate(linear);
        if (activation == Activation.TANH) {
            return raw >= 0.0 ? 1 : 0;
        }
        return raw >= 0.5 ? 1 : 0;
    }

    public double accuracy(double[][] x, int[] y) {
        int ok = 0;
        for (int i = 0; i < x.length; i++) {
            if (predict(x[i]) == y[i]) {
                ok++;
            }
        }
        return (double) ok / x.length;
    }

    private double activate(double x) {
        switch (activation) {
            case LINEAL:
                return x;
            case ESCALON:
                return x >= 0.0 ? 1.0 : 0.0;
            case SIGMOIDAL:
                return 1.0 / (1.0 + Math.exp(-x));
            case RELU:
                return Math.max(0.0, x);
            case SOFTMAX:
                return softmaxBinary(x);
            case TANH:
                return Math.tanh(x);
            default:
                return x;
        }
    }

    private double derivative(double x) {
        switch (activation) {
            case LINEAL:
                return 1.0;
            case ESCALON:
                return 1.0;
            case SIGMOIDAL:
                double s = 1.0 / (1.0 + Math.exp(-x));
                return s * (1.0 - s);
            case RELU:
                return x > 0.0 ? 1.0 : 0.0;
            case SOFTMAX:
                double sm = softmaxBinary(x);
                return sm * (1.0 - sm);
            case TANH:
                double t = Math.tanh(x);
                return 1.0 - t * t;
            default:
                return 1.0;
        }
    }

    private double softmaxBinary(double x) {
        double ex = Math.exp(x);
        return ex / (ex + 1.0);
    }

    private static double dot(double[] a, double[] b) {
        double sum = 0.0;
        for (int i = 0; i < a.length; i++) {
            sum += a[i] * b[i];
        }
        return sum;
    }
}

File: implementaciones/csharp/Perceptron.cs
using System;

public enum ActivationType
{
    Lineal,
    Escalon,
    Sigmoidal,
    Relu,
    Softmax,
    Tanh
}

public class Perceptron
{
    private readonly double learningRate;
    private readonly int epochs;
    private readonly ActivationType activation;
    private double[] weights = Array.Empty<double>();
    private double bias;

    public Perceptron(double learningRate = 0.1, int epochs = 50, ActivationType activation = ActivationType.Escalon)
    {
        if (epochs < 10)
        {
            throw new ArgumentException("epochs must be >= 10");
        }

        this.learningRate = learningRate;
        this.epochs = epochs;
        this.activation = activation;
    }

    public void Train(double[][] x, int[] y)
    {
        int samples = x.Length;
        int features = x[0].Length;
        weights = new double[features];
        bias = 0.0;

        for (int epoch = 0; epoch < epochs; epoch++)
        {
            for (int i = 0; i < samples; i++)
            {
                double linear = Dot(x[i], weights) + bias;
                double prediction = Activate(linear);
                double error = y[i] - prediction;
                double adjust = activation == ActivationType.Escalon
                    ? learningRate * error
                    : learningRate * error * Derivative(linear);

                for (int j = 0; j < features; j++)
                {
                    weights[j] += adjust * x[i][j];
                }

                bias += adjust;
            }
        }
    }

    public int Predict(double[] input)
    {
        double linear = Dot(input, weights) + bias;
        double raw = Activate(linear);
        return activation == ActivationType.Tanh
            ? (raw >= 0.0 ? 1 : 0)
            : (raw >= 0.5 ? 1 : 0);
    }

    public double Accuracy(double[][] x, int[] y)
    {
        int ok = 0;
        for (int i = 0; i < x.Length; i++)
        {
            if (Predict(x[i]) == y[i])
            {
                ok++;
            }
        }

        return (double)ok / x.Length;
    }

    private double Activate(double x)
    {
        return activation switch
        {
            ActivationType.Lineal => x,
            ActivationType.Escalon => x >= 0.0 ? 1.0 : 0.0,
            ActivationType.Sigmoidal => 1.0 / (1.0 + Math.Exp(-x)),
            ActivationType.Relu => Math.Max(0.0, x),
            ActivationType.Softmax => SoftmaxBinary(x),
            ActivationType.Tanh => Math.Tanh(x),
            _ => x
        };
    }

    private double Derivative(double x)
    {
        return activation switch
        {
            ActivationType.Lineal => 1.0,
            ActivationType.Escalon => 1.0,
            ActivationType.Sigmoidal => Activate(x) * (1.0 - Activate(x)),
            ActivationType.Relu => x > 0.0 ? 1.0 : 0.0,
            ActivationType.Softmax => SoftmaxBinary(x) * (1.0 - SoftmaxBinary(x)),
            ActivationType.Tanh => 1.0 - Math.Pow(Math.Tanh(x), 2),
            _ => 1.0
        };
    }

    // Softmax scalar simplificado para clasificacion binaria.
    private static double SoftmaxBinary(double x)
    {
        double ex = Math.Exp(x);
        return ex / (ex + 1.0);
    }

    private static double Dot(double[] a, double[] b)
    {
        double sum = 0.0;
        for (int i = 0; i < a.Length; i++)
        {
            sum += a[i] * b[i];
        }

        return sum;
    }
}
